File: src/src/ajshin/level_2/뒤에있는큰수찾기.java
import java.util.ArrayList;
import java.util.List;
import java.util.Stack;

public class 뒤에있는큰수찾기 {

  public List<Integer> solution(int[] numbers) {
    Stack<Integer> stack = new Stack<>();
    List<Integer> result = new ArrayList<>();
    for (int i = numbers.length-1; i >=0; i--) {
      int current = numbers[i];

      while (!stack.isEmpty() && stack.peek() <= current) {
        stack.pop();
      }

      if (stack.isEmpty()) {
        result.add(-1);
      } else {
        result.add(stack.peek());
      }
      stack.push(current);
    }
    List<Integer> reversedResult = new ArrayList<>();
    for (int i = result.size() - 1; i >= 0; i--) {
      reversedResult.add(result.get(i));
    }
    return reversedResult;
  }

  public static void main(String[] args) {
    뒤에있는큰수찾기 main = new 뒤에있는큰수찾기();
    int[] arr = {2, 3, 3, 5};
    for (int i : main.solution(arr)) {
      System.out.print(i);
    }
  }
}

File: src/src/yjhwang/level_2/뒤에있는큰수찾기.js
function solution(numbers) {
  const result = new Array(numbers.length).fill(-1);
  for (let i = numbers.length - 2; i >= 0; i--) {
    const currentNumber = numbers[i];
    for (let j = i + 1; j < numbers.length; j++) {
      const compareNumber = numbers[j];
      if (currentNumber < compareNumber) {
        result[i] = compareNumber;
        break;
      } else {
        if (result[j] === -1) {
          result[i] = -1;
          break;
        } else if (currentNumber < result[j]) {
          result[i] = result[j];
          break;
        }
      }
    }
  }

  return result;
}

console.log(solution([2, 3, 3, 5]));
console.log(solution([9, 1, 5, 3, 6, 2]));
